### PR TITLE
Fix #519: Make CSize an unsigned type

### LINKED ISSUE #368: CSignedSize missing
`package.scala` defines `CSize` (`size_t`), but doesn't have any equivalent for `ssize_t`. Would it be possible to add `CSignedSize`?

### LINKED ISSUE #519: CSize should be unsigned
My language lawyer license has long expired, so please forgive
the waste of valuable devo time if I am just muddled here.

In scala-native-git/nativelib/src/main/scala/scala/scalanative/native.package CSize
is defined as type Word, which is defined as type Long, which is signed.
```
type CSize = Word

 /** Int on 32-bit architectures and Long on 64-bit ones. */
type Word = Long
```
if CSize is supposed to be a strict analog for the C language  size_t,
shouldn't the underlying type be UWord, which is a ULong?
ssize_t is the signed variant.

I do not have a test case, but I would expect a test case such 
as (strlen(longString) > 0) to fail when the string had a length
greater than 32K.  Since size_t is expected, by the standard,
to hold at least 64K, this could probably happen.

The ULong to Long, unsigned to signed issue, probably gets
handled somewhere further down in the internals. Changing
to unsigned would preserve intended types to at least this level,
increasing readability/traceability/believability.

Thank you.
Lee

### LINKED ISSUE #732: CSize should be unsigned
It looks like `CSize` should be set to `UWord`, not `Word`, to align with the C and POSIX definitions of `size_t`:
https://github.com/scala-native/scala-native/blob/38fb182049698bf67df5be82ad63a4a3b2817315/nativelib/src/main/scala/scala/scalanative/native/package.scala#L72

This would be a good chance to also introduce a `type CSignedSize = Word` that corresponds to `ssize_t`

Reference:
http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_types.h.html

### PR DESCRIPTION
**Pull request by WojciechMazur**
For long time `CSize` was an alias to `Word` / `Long` however it was wrong behaviour. As @densh described in one of issues it should be considered as a bug and fixed with some breaking release. 

This PR fixes this bug by making `CSize` alias for `CWord` / `ULong`

Resolves #732 
Resolves #368 
Resolves #519

### CODE REVIEW

**Comment by WojciechMazur** [comment]
CI failed with single test due to coursier connectivity problem, when re-run it should pass and give green light

**Comment by ekrich** [comment]
I restarted that one - hopefully it will finish.

**Comment by LeeTibbert** [comment]
re: 128U  @ekrich you probably remember better than I...  I seem to remember reading in the code,  probably for Unsigned,
that forcing the .toUInt was a considered and deliberate design choice.  I forget the reason and took it on faith.  Of course,
that was then and this is now.

**Comment by LeeTibbert** [comment]
@WojciechMazur  Thank you mucho for fixing CSize. Nice to see a long standing design feature re-worked and fixed.

I do hope this PR gets a __prominent__ release note, especially the part about passing signed number to alloc/stackalloc 
(& kin). One of the key features of SN has been C interoperability and alloc/stackalloc is, as you know, extensively used there.
Deprecation is far better than a compile failure, but those are the people we do not want to astonish in a bad way. 
A TL;DR heads-up would go a long way to preserving the good will of people who are already kindly disposed towards
SN.   Perhaps you are way ahead of me on this.

**Comment by LeeTibbert** [comment]
I believe that this also allows my Issue #519 to be closed.

**Comment by ekrich** [comment]
@LeeTibbert I think that was a joke as I don't think we have unsigned literals but it would be pretty.

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package java.lang
 
 import scalanative.unsafe._
+import scalanative.unsigned._
 import scalanative.libc.errno
 
 import scalanative.posix.errno.ERANGE
```
[the remark this answers is not in the document]
Comment (ekrich) [reply javalib/src/main/scala/java/lang/IEEE754Helpers.scala:7]: Come to think about it, maybe this is better for another PR.

[the remark this answers is not in the document]
Comment (WojciechMazur) [reply javalib/src/main/scala/java/lang/IEEE754Helpers.scala:7]: I would agree that it might make cleaner commit history if it that change would be applied in other PR. Especially if it would be needed to make some additional changes or change dependencies in `build.sbt`

**Comment by sjrd** [comment]
This needs a rebase (and a regeneration of `Tag.scala` and perhaps others) now that the `CFuncPtr` PR was merged.

File: unit-tests/src/test/scala/scala/scalanative/unsafe/PtrBoxingTest.scala [hunk 2]
```diff
@@ -198,13 +198,14 @@ class PtrBoxingTest {
...
       val expectedInt    = 42
-      val expectedLength = 12
+      val expectedLength = 12.toULong
 
       assertEquals(expectedInt, x._1.apply())
       assertEquals(expectedInt, loadedGetInt())
 
-      assertEquals(expectedLength, x._2.apply(testStr))
-      assertEquals(expectedLength, loadedStringLength(testStr))
+      //assertEquals uses Object.equals which would return false in following cases
```
Comment (sjrd) [inline unit-tests/src/test/scala/scala/scalanative/unsafe/PtrBoxingTest.scala:206]: That seems very problematic. Two equivalent `ULong`s should compare equal according to `Object.equals`. It seems we should override `equals` and `hashCode` in all the `U` types. It should be in a separate PR, so please file an issue and reference from here (as well as from the other places where `assertEquals` was replaced by `assert`).

**Review by sjrd — CHANGES_REQUESTED** [review]

File: unit-tests/src/test/scala/scala/scalanative/unsafe/PtrBoxingTest.scala (hunk 2 above)
Comment (sjrd) [reply unit-tests/src/test/scala/scala/scalanative/unsafe/PtrBoxingTest.scala:206]: I guess this PR can now be rebased and amended to go back to using `assertEquals`, now that #2075 was merged.

**Review by sjrd — APPROVED** [review]
